src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,407 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/relax/transform/gradient.cc
+ * \brief A reverse-mode automatic differentiation. For a given function specified by the input
+ * global var, it generates a new function which computes the adjoints of the given function's
+ * inputs in respect of the function's output. Use require_grads to specify inputs which need to be
+ * differentiated instead of returning adjoints of all inputs.
+ *
+ * Now only supports differentiating a function in the IRModule with one dataflow block
+ * with respect to the only return value of the function. It needs to be scalar.
+ *
+ * For examples, see the MLP examples in tests/python/relax/test_transform_gradient.py and
+ * tests/python/relax/test_transform_gradient_numeric.py.
+ */
+
+#include <tvm/relax/expr_functor.h>
+#include <tvm/relax/op_attr_types.h>
+#include <tvm/relax/transform.h>
+
+#include <unordered_set>
+
+namespace tvm {
+namespace relax {
+
+class GradientMutator : public ExprMutator {
+ public:
+  explicit GradientMutator(IRModule mod, const Array<Var>& require_grads)
+      : ExprMutator(std::move(mod)), require_grads_(std::move(require_grads)) {}
+
+  Function FuncTransform(Function func) {
+    ICHECK(func->body->IsInstance<SeqExprNode>());
+
+    Array<Var> new_params;
+    for (Var param : func->params) {
+      Var new_param = Var(param->vid, param->shape(), param->checked_type(), param->span);
+      this->var_remap_[param->vid] = new_param;
+      new_params.push_back(new_param);
+    }
+    Expr new_body = VisitWithNewScope(func->body);
+
+    const SeqExprNode* seq_expr = new_body.as<SeqExprNode>();
+    // only support a single dataflow block
+    ICHECK(seq_expr->blocks.size() == 1);
+    ICHECK(seq_expr->blocks[0]->IsInstance<DataflowBlockNode>());
+    const DataflowBlockNode* block = seq_expr->blocks[0].as<DataflowBlockNode>();
+
+    builder_->BeginDataflowBlock();
+    for (const auto& binding : block->bindings) {
+      VisitBinding(binding);
+    }
+
+    // create adjoint var for inputs
+    for (size_t i = 0; i < new_params.size(); ++i) {
+      if (require_grads_.empty() || std::find(require_grads_.begin(), require_grads_.end(),
+                                              func->params[i]) != require_grads_.end()) {
+        CreateAdjointVar(new_params[i], false);
+      } else {
+        CreateAdjointVar(new_params[i], true);
+      }
+    }
+
+    // the return value of the function will be treated as target
+    if (const auto* node = seq_expr->body.as<VarNode>()) {
+      const Var& target = GetRef<Var>(node);
+      CheckTarget(target);
+      target_var_ = target;
+    } else {
+      LOG(FATAL) << "the body of the function is not a relax.Var";
+    }
+
+    // reverse-mode ad
+    for (int i = block->bindings.size() - 1; i >= 0; --i) {
```

Comment:
This is dangerous (or actually erroneous) since whenever `block->bindings` is empty, `block->bindings.size() - 1` will overflow to the maximum value of `size_t`.

Thus I would suggesting always do static cast at such places.
```suggestion
    for (int i = static_cast<int>(block->bindings.size()) - 1; i >= 0; --i) {
```

Reply:
```c++
// test.cpp
#include <cstddef>

int main() {
  size_t a = 0;
  int b = a - 1;
  return b;
}
```
```shell
>>> g++ -o test test.cpp
>>> ./test
>>> echo $?
255
```

Reply:
wow you have a good insight about bugs! I never think of this possibility

Reply:
You can be the same by always being alert to `size_t` :-)